Services/User/Profile/classes/class.ilPersonalProfileGUI.php

```diff
@@ -527,7 +533,124 @@ protected function showUserAgreement()
         $this->tpl->setPermanentLink('usr', null, 'agreement');
         $this->tpl->printToStdout();
     }
-    
+
+    protected function showConsentWithdrawalConfirmation() : void
+    {
+        $this->tabs->clearTargets();
+        $this->tabs->clearSubTabs();
+        $this->tpl->setTitle($this->lng->txt('withdraw_consent'));
+
+        $defaultAuth = AUTH_LOCAL;
+        if ($this->setting->get('auth_mode')) {
+            $defaultAuth = $this->setting->get('auth_mode');
+        }
+
+        $tpl = new \ilTemplate('tpl.withdraw_terms_of_service.html', true, true, 'Services/TermsOfService');
+        /** @var ilObjUser $user */
+        $user = $GLOBALS['DIC']->user();
+        if ( $user->getAuthMode() == AUTH_LDAP
+            || ( $user->getAuthMode() == 'default' && $defaultAuth == AUTH_LDAP) )
+        {
+            $message = nl2br(
+                $this->lng->txt('withdrawal_mail_info')
+                . $this->lng->txt('withdrawal_mail_text')
+            );
+
+            $tpl->setVariable('TERMS_OF_SERVICE_WITHDRAWAL_CONTENT', $message);
+            $tpl->setVariable('FORM_ACTION', $this->ctrl->getFormAction($this,'cmd[withdrawAcceptanceLDAP]'));
+        } else {
+            $tpl->setVariable('TERMS_OF_SERVICE_WITHDRAWAL_CONTENT', $this->lng->txt('withdraw_consent_info'));
+            $tpl->setVariable('FORM_ACTION', $this->ctrl->getFormAction($this,'cmd[withdrawAcceptance]'));
+        }
+
+        $tpl->setVariable('WITHDRAW_TERMS_OF_SERVICE', $this->lng->txt('withdraw_usr_agreement'));
+        $tpl->setVariable('TXT_WITHDRAW', $this->lng->txt('withdraw'));
+        $tpl->setVariable('TXT_CANCEL', $this->lng->txt('cancel'));
+
+        $this->tpl->setContent($tpl->get());
+        $this->tpl->setPermanentLink('usr', null, 'agreement');
+        $this->tpl->printToStdout();
+    }
+
+    protected function cancelWithdrawal() : void
+    {
+        $this->user->writePref('consent_withdrawal_requested', 0);
+        ilInitialisation::redirectToStartingPage();
+    }
+
+    protected function withdrawAcceptance() : void
+    {
+        $helper = new \ilTermsOfServiceHelper();
+        $helper->resetAcceptance($this->user);
+
+        $shouldDeleteAccountOnWithdrawal = $this->setting->get(
+            'tos_withdrawal_usr_deletion',
+            false
+        );
+
+        if($shouldDeleteAccountOnWithdrawal == 1 &&
+            ($this->user->getAuthMode() == AUTH_LOCAL || $this->user->getAuthMode() == 'default')
+        ) {
+            $accdel = 1;
+        } else {
+            $accdel = 0;
+        }
+
+        $domainEvent = new ilTermsOfServiceEventWithdrawn($this->user->getId());
+        $this->eventHandler->raise(
+            'Services/TermsOfService',
+            'ilTermsOfServiceEventWithdrawn',
+            ['event' => $domainEvent]
+        );
+
+        ilSession::setClosingContext(ilSession::SESSION_CLOSE_USER);
+        $GLOBALS['DIC']['ilAuthSession']->logout();
```

Comment:
Please don't use the `DIC` in instance methods. Assign `$GLOBALS['DIC']['ilAuthSession']` to a member variable in the constructor.